pkg/sql/catalog/descpb/structured.proto

```diff
@@ -1440,7 +1442,21 @@ message TypeDescriptor {
   // descriptor being changed as part of a declarative schema change.
   optional cockroach.sql.schemachanger.scpb.DescriptorState declarative_schema_changer_state = 17;
 
-  // Next field is 18.
+  message Composite {
+    option (gogoproto.equal) = true;
+    // TupleContents returns a slice containing the type of each tuple field. This
+    // is nil for non-TUPLE types.
+    repeated sql.sem.types.T tuple_contents = 1;
+
+    // TupleLabels returns a slice containing the labels of each tuple field. This
+    // is nil for non-TUPLE types, or if the TUPLE type does not specify labels.
+    repeated string tuple_labels = 2;
+  }
+
+  // composite is the list of fields if this is a composite type.
+  optional Composite composite = 18;
```

Comment:
Rather than `Composite` containing slices of metadata which all need to be of the same length, would it be preferable to have `Composite` contain a `repeated CompositeElement` the latter being a sum type for each tuple element?

Right now it doesn't matter very much but at some point we're going to want to support user-defined types and each element is going to have more metadata like an optional forward-reference descriptor ID, etc. So I'm keen for this to evolve gracefully.